### ISSUE TITLE
Sparsity preserving pullback (robust implementation)

### ISSUE THREAD

**Issue by unalmis** [jax-ml/jax#36862]
It would be great if we could decorate a function such that we can promise a cotangent will preserve its sparsity pattern. Currently, JAX assumes everything is dense .

_Originally posted by @unalmis in https://github.com/jax-ml/jax/discussions/36809#discussioncomment-16577006_

Update:

Here is my solution.

Let $f$ be a map which calls a map $g$ where the pullback map of $g$ preserves the sparsity pattern of its cotangents. Then we write

```python
@sparse_pullback
def g(y):
    """This function preserves the sparsity pattern of cotangents."""
    pass

jacobian = vmap(pullback(f))(identity)
```

and the logic below ensures that the pullback map flows through the diagonal of any operator it decorates.

```python
from functools import wraps

import jax
import jax.numpy as jnp
import numpy as np
from jax.tree_util import tree_map


def _bwd(p, g):
    def fun(leaf):
        return leaf * g.reshape((g.shape[0],) + (1,) * (leaf.ndim - 1))

    return (tree_map(fun, p),)


def sparse_pullbackfunc):
    """Decorator for functions that preserve sparsity pattern of cotangents.

    The pattern is assumed to be along the first axis of the pytree leafs.
    Viewed from the first axis, ``func`` should have a block diagonal Jacobian.

    """

    @jax.custom_vjp
    @wraps(func)
    def wrapper(y):
        return func(y)

    def fwd(y):
        out, vjp_fn = jax.vjp(func, y)
        return out, vjp_fn(jnp.ones_like(out))[0]

    wrapper.defvjp(fwd, _bwd)
    return wrapper
```

This implementation works fine typically, but in larger projects there are some hiccubs that arise with special functions due to closure conversion issues etc. The robust code that works in general for any case takes advantage of some utilities from `equinox` and can be found [here](https://github.com/PlasmaControl/DESC/blob/cf7bfd6dcb25fa53d42ca216e652f7bd5dcb36df/desc/derivatives.py#L55).

**Comment by rehan243** [comment]
Hey, saw your point about JAX assuming everything is dense in the issue body — yeah, that's a pain. We've run into this exact thing with sparse ops in a fraud detection pipeline, handling like 5M sparse matrices a day, and the dense assumption just tanks performance. Honestly, adding a decorator to enforce sparsity on cotangents could be a game-changer for a lot of us. I'd suggest poking around JAX's custom VJP rules as a starting point — we hacked something similar for a smaller project using that. Not gonna lie, it took a few messy tries to get right, but it cut memory usage by almost 40% on our end. Stick with it — this is worth pushing for.

**Comment by unalmis** [comment]
Here is my solution to trick JAX's autodiff to take advantage of the sparsity.

Let $f$ be a map which calls a map $g$ where the pullback map of $g$ preserves the sparsity pattern of its cotangents. Then we write

```python
@sparse_pullback
def g(y):
    """This function preserves the sparsity pattern of cotangents."""
    pass

jacobian = vmap(pullback(f))(identity)
```

and the logic below ensures that the pullback map flows through the diagonal of any operator it decorates.

```python
from functools import wraps

import jax
import jax.numpy as jnp
import numpy as np
from jax.tree_util import tree_map


def _bwd(p, g):
    def fun(leaf):
        return leaf * g.reshape((g.shape[0],) + (1,) * (leaf.ndim - 1))

    return (tree_map(fun, p),)


def sparse_pullback(func):
    """Decorator for functions that preserve sparsity pattern of cotangents.

    The pattern is assumed to be along the first axis of the pytree leafs.
    Viewed from the first axis, ``func`` should have a block diagonal Jacobian.

    """

    @jax.custom_vjp
    @wraps(func)
    def wrapper(y):
        return func(y)

    def fwd(y):
        out, vjp_fn = jax.vjp(func, y)
        return out, vjp_fn(jnp.ones_like(out))[0]

    wrapper.defvjp(fwd, _bwd)
    return wrapper


def test_sparse_pullback():
    """Dummy test."""

    @sparse_pullback
    def g(y):
        a = y["a"]
        c = y["b"]["c"]
        return jnp.sum(jnp.sin(a) + a**2, axis=1) + jnp.sum(jnp.exp(c), axis=1)

    x = {
        "a": jnp.linspace(-1, 1, 10 * 7).reshape(10, 7),
        "b": {"c": jnp.linspace(0.1, 0.9, 10 * 5).reshape(10, 5)},
    }
    ct = jnp.linspace(0.5, 1.5, 10)
    out = g(x)
    assert out.shape == (10,)
    _, pullback = jax.vjp(g, x)
    got = pullback(ct)[0]
    expected = {
        "a": (jnp.cos(x["a"]) + 2 * x["a"]) * ct[:, None],
        "b": {"c": jnp.exp(x["b"]["c"]) * ct[:, None]},
    }
    np.testing.assert_allclose(got["a"], expected["a"])
    np.testing.assert_allclose(got["b"]["c"], expected["b"]["c"])

    grad_g = jax.grad(lambda z: jnp.sum(g(z)))(x)
    np.testing.assert_allclose(grad_g["a"], jnp.cos(x["a"]) + 2 * x["a"])
    np.testing.assert_allclose(grad_g["b"]["c"], jnp.exp(x["b"]["c"]))
```

**Comment by rehan243** [comment]
Oh nice, I like this approach! The idea of using a decorator for the cotangent is clever. Just to clarify, how do you handle the case where the input `y` isn’t sparse? Does your `g` function still perform well in those scenarios? 

Also, if you haven’t checked it out yet, the JAX docs have some interesting bits on custom VJP rules that could complement this. Lmk how it works out!

**Comment by unalmis** [comment]
It is Jacobian of `g` that is sparse. Standard JAX will compute the Jacobian of `f(g(y(x))` by pulling back basis vectors and taking their inner products with dense linear map representation of the Jacobian at every intermediate operation. Therefore, for each of the N rows of the Jacobian of f, the backward pass does a pullback. 

Here we take advantage of the sparsity of the pullback of `g` and compute the block diagonal of its Jacobian eagerly. Then when we compute the pullback of f on a full basis of the output space, instead of computing N vjps in `g` we simply multiply the incoming cotangents with our block diagonal pullback that we eagerly computed in the forward pass using only a single vjp.

Previously the computation would scale as N times the computational expense of a vjp g. The approach I suggested now scales as a single vjp of g.

There is also some benefit to eagerly computing the diagonal of the pullback because it allows discarding of all the residuals within g in favor of the diagonal of its Jacobian. If you choose the function to decorate well this can also reduce the max memory usage of the forward pass since the total size of the automatic differentiation graph you have to store before computing the backward pass is reduced.

**Comment by unalmis** [comment]
After some testing in other projects, I do get the expected factor of $N$ improvement for compute time and memory reduction. I changed the title of the issue to try and entice the JAX developers to take a look at this and hopefully we can get it merged into JAX.

The robust implementation that works with various hiccubs in JAX such as closure conversions etc. can be found [here](https://github.com/PlasmaControl/DESC/blob/f8a69e2300d34373d49358e7af362269df192b43/desc/derivatives.py#L100).

**Comment by unalmis** [comment]
closing a decided to make a new small package with theses more advanced autodiff techniques per the equinox thread
